lib/miq_expression.rb

```diff
@@ -323,31 +335,124 @@ def preprocess_exp!(exp)
+    end
+    [filtered_children, seen]
+  end
+  private :reduce_exp_children
+
+  # private
+  # remove nodes from the ast that we don't want to process in the given mode
+  #
+  # We in essence are cutting up the expression into 2 sub expressions that
+  # are applied sequentially:
+  #
+  # exp == (sql_exp) AND (ruby_exp)
+  #
+  # Although we could apply the full expression in ruby, it is more efficient
+  # to use the sql friendly portion of the expression in the database to reduce
+  # the number of records returned from the database.
+  #
+  # Due to logic reasons, sometimes we still apply sql friendly expressions in ruby.
+  #
+  # @param exp  [Hash]               ast for miq_expression
+  # @param mode [MODE_RUBY|MODE_SQL] whether we are reducing for sql or ruby generation
+  # @param swap [boolean]            true if we are in a NOT clause and applying Demorgan's law
+  # @returns [Hash, mode]
+  #   Hash: contains ast for children that are approperiate for the given mode
+  #   mode: the children whether they contain ruby, sql, or both
+  #
+  # Implementation notes (more of a logic tutorial)
+  #
+  # From here we determine the results for a compond mode:
+  #
+  # exp          |==>| exp (mode=sql) |and| exp (mode=ruby)
+  # -------------|---|----------------|---|-----------------
+  # sql  OR sql  |==>| sql OR sql     |AND|
+  # sql  OR ruby |==>|                |AND| sql OR ruby
+  # ruby OR ruby |==>|                |AND| ruby OR ruby
+  #
+  # The OR clause shows an example where sql friendly filters are applied in ruby
+  #
+  # The all sql or all ruby results line up with the filtered child nodes.
+  # When there are both types of children, the result lines up with either an all or nothing approach
+  #
+  #
+  # exp             |==>| exp (mode=sql) |and| exp (mode=ruby)
+  # ----------------|---|----------------|---|-----------------
+  # sql1  AND sql2  |==>| sql1 AND sql2  |AND|
+  # sql   AND ruby  |==>| sql            |AND| ruby
+  # ruby1 AND ruby2 |==>|                |AND| ruby1 AND ruby2
+  #
+  # In the AND case, the sql nodes line up with the filtered child nodes
+  #
+  #
+  # exp                |==>| exp (mode=sql) |and| exp (mode=ruby)
+  # -------------------|---|----------------|---|-----------------
+  # !(sql  OR sql)     |==>| !(sql OR sql)  |AND|
+  # !(sql  OR ruby)    |==>| !(sql)         |AND| !(ruby)
+  # !(ruby1 OR ruby2)  |==>|                |AND| !(ruby1 OR ruby2)
+  #
+  # exp                |==>| exp (mode=sql) |and| exp (mode=ruby)
+  # -------------------|---|----------------|---|-----------------
+  # !(sql   AND sql)   |==>| !(sql AND sql) |AND|
+  # !(sql   AND ruby)  |==>|                |AND| !(sql AND ruby)
+  # !(ruby1 AND ruby2) |==>|                |AND| !(ruby1 AND ruby2)
+  #
+  # Inside a NOT, the OR acts like the AND, and the AND acts like the OR
+  # so swap is introduced to handle Demorgan's law
+  #
+  def reduce_exp(exp, mode, swap: false)
```

Comment:
Is reduce_exp modifying the original exp, or just returning a copy?   If the former, can we name it `reduce_exp!`

Reply:
original expression is not changed, but it does point to some of the entries from it

Reply:
Sorry i'm late to this, but seems like we're dividing/filtering based on the mode (ruby/sql).  `reduce` makes me think of the ruby enumerable method and that we're just simplifying. Partitioning by mode is what I think this is doing after reviewing the code.  It's too bad partition is also an enumerable method.  I guess divide is the best I can offer.

Reply:
I will rename to `prune_exp`. We are pruning out the nodes that are no relevant